File: lingting-spring-redis/src/main/java/live/lingting/spring/redis/prefix/StringKeyPrefixSerializer.java
package live.lingting.spring.redis.prefix;

import org.springframework.data.redis.serializer.StringRedisSerializer;

/**
 * @author lingting 2024-04-17 16:05
 */
public class StringKeyPrefixSerializer extends StringRedisSerializer {

	private final KeyPrefixConvert convert;

	public StringKeyPrefixSerializer(KeyPrefixConvert convert) {
		this.convert = convert;
	}

	@Override
	public byte[] serialize(String value) {
		byte[] bytes = super.serialize(value);
		return convert.isMatch(bytes) ? convert.wrap(bytes) : bytes;
	}

	@Override
	public String deserialize(byte[] bytes) {
		byte[] target = convert.isMatch(bytes) ? convert.unwrap(bytes) : bytes;
		return super.deserialize(target);
	}

}

File: lingting-spring-redis/src/main/kotlin/live/lingting/spring/redis/prefix/StringKeyPrefixSerializer.kt
package live.lingting.spring.redis.prefix

import org.springframework.data.redis.serializer.StringRedisSerializer

/**
 * @author lingting 2024-04-17 16:05
 */
@Suppress("WRONG_NULLABILITY_FOR_JAVA_OVERRIDE")
class StringKeyPrefixSerializer(private val convert: KeyPrefixConvert) : StringRedisSerializer() {
    override fun serialize(value: String): ByteArray {
        val bytes = super.serialize(value)
        return if (convert.isMatch(bytes)) convert.wrap(bytes) else bytes
    }

    override fun deserialize(bytes: ByteArray): String {
        val target = if (convert.isMatch(bytes)) convert.unwrap(bytes) else bytes
        return super.deserialize(target)
    }
}
